File: skills/slide-gen-agent/scripts/pdfExporter.ts
import * as fs from 'fs';
import * as path from 'path';

/**
 * Gathers all slide PNGs in the session slides directory and compiles them into a single PDF presentation.
 * @param sessionPath The absolute session path
 * @param pdfFileName Optional custom filename for the output PDF (defaults to presentation.pdf)
 * @param tool_context Optional ADK context to register the artifact
 */
export async function exportSessionToPdf(sessionPath: string, pdfFileName?: string, tool_context?: any): Promise<{ message: string; pdfName: string; pdfPath: string }> {
  // @ts-ignore
  const { PDFDocument } = await import('pdf-lib'); // Dynamic import to keep startup fast
  // Read all files in the session root directory and filter for slide_xx.png files
  const files = fs.readdirSync(sessionPath);
  const pngFiles = files
    .filter(file => /^slide_\d+\.png$/.test(file))
    .sort(); // Sorts numerically (slide_01.png, slide_02.png, ...)

  if (pngFiles.length === 0) {
    throw new Error(`No slide PNG images found in ${sessionPath}. Make sure to generate slide images first.`);
  }

  const pdfDoc = await PDFDocument.create();

  for (const fileName of pngFiles) {
    const filePath = path.join(sessionPath, fileName);
    const pngImageBytes = fs.readFileSync(filePath);
    const pngImage = await pdfDoc.embedPng(pngImageBytes);

    // Embed each PNG image page with its native dimensions (typically 16:9 widescreen)
    const page = pdfDoc.addPage([pngImage.width, pngImage.height]);
    page.drawImage(pngImage, {
      x: 0,
      y: 0,
      width: pngImage.width,
      height: pngImage.height,
    });
  }

  const pdfBytes = await pdfDoc.save();
  const outputName = pdfFileName ? (pdfFileName.endsWith('.pdf') ? pdfFileName : `${pdfFileName}.pdf`) : 'presentation.pdf';
  const outputPath = path.join(sessionPath, outputName);

  fs.writeFileSync(outputPath, pdfBytes);

  if (tool_context) {
    await tool_context.saveArtifact(outputName, {
      inlineData: {
        data: Buffer.from(pdfBytes).toString('base64'),
        mimeType: 'application/pdf',
      }
    });
  }

  return {
    message: `Successfully compiled ${pngFiles.length} slides into a single PDF presentation.`,
    pdfName: outputName,
    pdfPath: outputPath,
  };
}

// CLI Support (Allows running from Antigravity/Codex terminal directly)
if (typeof process !== 'undefined' && process.argv && process.argv[1] && (process.argv[1].endsWith('pdfExporter.ts') || process.argv[1].endsWith('pdfExporter.js'))) {
  const sessionPathArg = process.argv[2];
  const pdfNameArg = process.argv[3];
  if (!sessionPathArg) {
    console.error('Usage: node pdfExporter.js <sessionPath> [pdfFileName]');
    process.exit(1);
  }
  exportSessionToPdf(sessionPathArg, pdfNameArg)
    .then(result => {
      console.log(JSON.stringify(result, null, 2));
      process.exit(0);
    })
    .catch(err => {
      console.error('PDF compilation failed:', err.message);
      process.exit(1);
    });
}

File: skills/slide-gen-agent/scripts/pdf_exporter.py
import os
import sys
import glob
import json
from PIL import Image

def export_session_to_pdf(session_path: str, pdf_file_name: str = None) -> dict:
    """Gathers all slide PNGs and compiles them into a single PDF presentation.
    Supports both subfolder 'slides/' and root session directory structures.
    """
    # 1. Search in 'slides/' subfolder first
    slides_dir = os.path.join(session_path, 'slides')
    png_files = sorted(glob.glob(os.path.join(slides_dir, 'slide_*.png')))
    
    # 2. Fallback to session root directory if subfolder is empty
    if not png_files:
        slides_dir = session_path
        png_files = sorted(glob.glob(os.path.join(slides_dir, 'slide_*.png')))
        
    if not png_files:
        raise ValueError(f"No slide PNG images found in {session_path}. Make sure to generate slide images first.")
        
    # Open and convert to RGB
    images = [Image.open(f).convert('RGB') for f in png_files]
    
    # Resolve output name
    output_name = pdf_file_name or 'presentation.pdf'
    if not output_name.endswith('.pdf'):
        output_name += '.pdf'
        
    output_path = os.path.join(session_path, output_name)
    
    # Compile
    images[0].save(
        output_path,
        save_all=True,
        append_images=images[1:]
    )
    
    return {
        "message": f"Successfully compiled {len(png_files)} slides into a single PDF presentation.",
        "pdfName": output_name,
        "pdfPath": output_path
    }

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python3 pdf_exporter.py <sessionPath> [pdfFileName]", file=sys.stderr)
        sys.exit(1)
        
    session_path_arg = sys.argv[1]
    pdf_name_arg = sys.argv[2] if len(sys.argv) > 2 else None
    
    try:
        result = export_session_to_pdf(session_path_arg, pdf_name_arg)
        print(json.dumps(result, indent=2))
        sys.exit(0)
    except Exception as e:
        print(f"PDF compilation failed: {str(e)}", file=sys.stderr)
        sys.exit(1)
